### DISCUSSION TITLE
Facing issues with saving nifti image with SaveImaged because of multi channel input

### DISCUSSION THREAD

**Discussion by lenlobo** [Project-MONAI/MONAI#5335] [Q&A] [answered]
I am working with 3d segmentation using Monai and Nifti files as inputs and using a model which has 6 input channels, I get an error in the inference testing of the model.
```
class Net(pytorch_lightning.LightningModule):
    def __init__(self):
        super().__init__()
        self._model = UNet(
            dimensions=3,
            in_channels=6,
            out_channels=2,
            channels=(16, 32, 64, 128, 256),
            strides=(2, 2, 2, 2),
            num_res_units=2,
            norm=Norm.BATCH,
        )
```

So after using the tutorial code from "https://github.com/Project-MONAI/tutorials/blob/72728b678947bc21443d02d3a19d29f01c9eed01/3d_segmentation/spleen_segmentation_3d.ipynb", as a reference, the issue I end up is at post transforms to revert the pretransforms and save the output file.

As I am using 6 inputs, and end up with only one output, I get an error stating, 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [31], in <cell line: 1>()
      1 post_transforms = Compose([
----> 2     Invertd(
      3         keys="pred",
      4         transform=test_transforms,
      5         orig_keys=img_keys,
      6         meta_keys="pred_meta_dict",
      7         meta_key_postfix="meta_dict",
      8         nearest_interp=False,
      9         to_tensor=True,
     10     ),
     11     AsDiscreted(keys="pred", argmax=True, to_onehot=2),
     12     SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="./out", output_postfix="seg", resample=False),
     13 ])


File /opt/monai/monai/transforms/post/dictionary.py:592, in Invertd.__init__(self, keys, transform, orig_keys, meta_keys, orig_meta_keys, meta_key_postfix, nearest_interp, to_tensor, device, post_func, allow_missing_keys)
    590     raise ValueError("transform is not invertible, can't invert transform for the data.")
    591 self.transform = transform
--> 592 self.orig_keys = ensure_tuple_rep(orig_keys, len(self.keys)) if orig_keys is not None else self.keys
    593 self.meta_keys = ensure_tuple_rep(None, len(self.keys)) if meta_keys is None else ensure_tuple(meta_keys)
    594 if len(self.keys) != len(self.meta_keys):

File /opt/monai/monai/utils/misc.py:154, in ensure_tuple_rep(tup, dim)
    151 if len(tup) == dim:
    152     return tuple(tup)
--> 154 raise ValueError(f"Sequence must have length {dim}, got {len(tup)}.")

**ValueError: Sequence must have length 1, got 6.`**

```

Now, as I am passing img_keys which is a list of 6 keys, _eg, input_1, input_2, etc_, it requires one single key instead of 6, but in pre transform I have used 6 keys because of 6 input channels. 
This is what my output dictionary keys look like after sliding_window_inference : 
`
dict_keys(['input_1', 'input_2', 'input_3', 'input_4', 'input_5', 'input_6', 'input_1_meta_dict', 'input_2_meta_dict', 'input_3_meta_dict', 'input_4_meta_dict', 'input_5_meta_dict', 'input_6_meta_dict', 'pred'])
`
To summarize, I have 6 inputs, using all 6 inputs, I generate an output tensor, but when I try to save it using the post transform, I face issues with Value Error.
Any solution for this, multi input channel, single output channel?

**Comment by KumoLiu** [comment, marked as answer]
Hi @lenlobo, If you have images with 6 input channels, you can simply use one key to preprocess them like:
```
trans = Compose([
    LoadImaged(keys='image'),
    ......
])
```
```
Datalist like: [
{'image': [input_1, input_2, ....]
'label': label},
......
```
]
You can also refer to [brats segmentation tutorial](https://github.com/Project-MONAI/tutorials/blob/72728b678947bc21443d02d3a19d29f01c9eed01/3d_segmentation/brats_segmentation_3d.ipynb) for reference which uses 4 modalities as input.
Hope it can help you, thanks!

**Reply by lenlobo** [reply]
Hi, thanks for the reply, this makes sense to me, but how should I define the datalist as? A bit more of code or a reference maybe would help me a lot! Meanwhile I would go through brats tutorial :)

**Reply by KumoLiu** [reply]
Hi, @lenlobo You need to define a list containing a series `dict` of your input data, like:
```
[
{
  'image': [
    "/your_path1/img1.nii", 
    "/your_path1/img2.nii",
    "/your_path1/img3.nii"
],
  'label': "/your_path1/label.nii"
},
{
  'image': [
    "/your_path2/img1.nii", 
    "/your_path2/img2.nii",
    "/your_path2/img3.nii"
],
  'label': "/your_path2/label.nii"
}, 
...,
]

```
Hope this code sample could help.

**Reply by lenlobo** [reply]
Hi, @KumoLiu I did that and thought it would work, but during inference particulary when I try to load data using 
`
for test_data in enumerate(test_loader):
.
.
.
 ` I got a new RuntimeError, 

```
RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/opt/monai/monai/transforms/transform.py", line 90, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "/opt/monai/monai/transforms/transform.py", line 54, in _apply_transform
    return transform(parameters)
  File "/opt/monai/monai/transforms/spatial/dictionary.py", line 439, in __call__
    d[key] = self.ornt_transform(d[key])
  File "/opt/monai/monai/transforms/spatial/array.py", line 646, in __call__
    raise ValueError(
ValueError: axcodes must match data_array spatially, got axcodes=3D data_array=4D

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/monai/monai/transforms/transform.py", line 90, in apply_transform
    return _apply_transform(transform, data, unpack_items)
  File "/opt/monai/monai/transforms/transform.py", line 54, in _apply_transform
    return transform(parameters)
  File "/opt/monai/monai/transforms/compose.py", line 173, in __call__
    input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
  File "/opt/monai/monai/transforms/transform.py", line 114, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.spatial.dictionary.Orientationd object at 0x7f22f0d1d5e0>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 300, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/monai/monai/data/dataset.py", line 97, in __getitem__
    return self._transform(index)
  File "/opt/monai/monai/data/dataset.py", line 821, in _transform
    return super()._transform(index_)
  File "/opt/monai/monai/data/dataset.py", line 83, in _transform
    return apply_transform(self.transform, data_i) if self.transform is not None else data_i
  File "/opt/monai/monai/transforms/transform.py", line 114, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <monai.transforms.compose.Compose object at 0x7f22f0d1d7c0>
```

As this is my company's project I can't share much information with you, but let me know if you need to know anything to help me here.
Thanks!

**Reply by KumoLiu** [reply]
Hi @lenlobo, from the error message it seems your data shape is not correct. Your data shape should be with shape CxHxWxD, maybe check your data shape. Hope it can help you, thanks!

**Reply by lenlobo** [reply]
Hey @KumoLiu , but it is correct, the way I was doing it till now, was by using torch.cat and passing all 6 tensors to it. It did give me the output but was not saving the output using SaveImaged because of the errors. I used nibable and tried to retrieve the affine and saved the image and got the output, but I do want to use SaveImaged because that seems way more convenient. 
So, I tried to pass 
```
test_files = [{
  'image': [
    'path/to/file/1',
    'path/to/file/2',
    'path/to/file/3',
    'path/to/file/4',
    'path/to/file/5',
    'path/to/file/6']
}]
``` 
as your suggestion and this generated the above error in my last reply. I also tried to do this
```
test_files = [{
    'image': [ 'path/to/file/1']
}]
```
which worked for me to read the file, but as my model expects 6 channels, it was not possible to generate an output. I don't really think, it's a shape issue, but the loader having issues with loading the dict-list series.

**Reply by KumoLiu** [reply]
Hi @lenlobo, the data from dataloader will have batch dim then the shape of the data will be BxCxHxWxD.

**Reply by lenlobo** [reply]
Hi @KumoLiu, that's correct, and I am having a shape of  BxCxHxWxD while passing it to the inference. My issue is to create a dict series with "image" key for multiple images, which isn't working as you had suggested.

**Reply by KumoLiu** [reply]
Hi @lenlobo, I'm a bit confused by your question, from your error message it looks like it's the shape that's the problem, you shouldn't pass data containing the batch dim to transform. Did you detach batch dim when you do post_transforms like this?
```
val_data = [post_transforms(i) for i in decollate_batch(val_data)]
```

**Reply by lenlobo** [reply]
Hi @KumoLiu , so I haven't yet gotten to that step, because I was just trying to print out the test data using this.
```
for test_data in enumerate(test_loader):
    print(test_data)
```
After this I'll pass the test data to the inference model to make the prediction, and then use `test_data = [post_transforms(i) for i in decollate_batch(test_data)]` but I haven't gotten here yet.

I was talking to my colleague about this and he mentioned that,  
```
test_files = [{
  'image': [
    'path/to/file/1',
    'path/to/file/2',
    'path/to/file/3',
    'path/to/file/4',
    'path/to/file/5',
    'path/to/file/6']
}]

``` 
with doing this, I am **stacking up 6 3d images which makes the input 4d** and messes up the shape, which makes sense because when I just use one file, it works. Mind you, if I didnt mention it before, all the 6 inputs are for a single image/sample, just gathered from 6 different sensors.

**Reply by KumoLiu** [reply]
Did you add channel dim in your transform? Since 6 will become the channel dim, you don't need to add channel dim anymore and just make sure the channel dim(6) is in the first dimension.

**Reply by lenlobo** [reply]
Hi @KumoLiu,
```
test_transforms= Compose(
                [
                    LoadImaged(keys="image"),
                    AddChanneld(keys="image"),
                    ScaleIntensityd(keys="image",minv=0.0, maxv=1.0),
                    Orientationd(keys="image", axcodes="RAS"),
                     ToTensord(keys="image")
                    
                ])
test_ds = CacheDataset(data=test_files, transform=test_transforms, cache_rate=0.01, num_workers=6)
#test_ds = Dataset(data=val_files, transform=test_transforms)
test_loader = torch.utils.data.DataLoader(test_ds, batch_size=1, num_workers=1)
```
This is my transform block of code,

**Reply by KumoLiu** [reply]
Hi @lenlobo, you can remove `AddChanneld(keys="image")` this line since 6 will be the channel dim. thanks!

**Reply by lenlobo** [reply]
Hi @KumoLiu , so I feel that worked and I could predict the output tensor, but I'm still having issues with post_transforms. I'll give you some more insights into my code.

My pre-transform,
```
test_transforms= Compose(
                [
                    LoadImaged(keys="image"),
                    
                    ScaleIntensityd(keys="image",minv=0.0, maxv=1.0),
                    Orientationd(keys="image", axcodes="RAS"),
                     ToTensord(keys="image")
                    
                ])
test_ds = CacheDataset(data=test_files, transform=test_transforms, cache_rate=0.01, num_workers=4)
test_loader = torch.utils.data.DataLoader(test_ds, batch_size=1, num_workers=1)
```
My post-transform,
```
post_transforms = Compose([
    Invertd(
        keys="pred",
        transform=test_transforms,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        meta_key_postfix="meta_dict",
        nearest_interp=False,
        to_tensor=True,
    ),
    AsDiscreted(keys="pred", argmax=True, to_onehot=2),
    SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="./out", output_postfix="seg", resample=False),
])
```
and, my inference code,
```
net.eval()
device = gpu_device
net.to(device)
#print("Starting Inference")
with torch.no_grad():
    for test_data in enumerate(test_loader):
        test_input = test_data['image']
        roi_size = (96, 96, 96)
        sw_batch_size = 32
        overlap = 0.5,
        print(i)
        test_data['pred'] = sliding_window_inference(
            test_input,
            roi_size, sw_batch_size, net, sw_device= device, device="cpu")      
        test_data = [post_transforms(i) for i in decollate_batch(test_data)]
```
When I run this, I get an error at the last line, and when I check test_data.keys() I get, `dict_keys(['image', 'image_meta_dict', 'pred'])`, which seems correct.
The output for my test_data['pred'] is just a tensor and unlike the 3d_segmentation tutorial, I do not have metadata information on the 'pred' key for some reason?
Now, on the last line the error that I get is,
```
KeyError                                  Traceback (most recent call last)
File /opt/monai/monai/transforms/transform.py:90, in apply_transform(transform, data, map_items, unpack_items, log_stats)
     89         return [_apply_transform(transform, item, unpack_items) for item in data]
---> 90     return _apply_transform(transform, data, unpack_items)
     91 except Exception as e:

File /opt/monai/monai/transforms/transform.py:54, in _apply_transform(transform, parameters, unpack_parameters)
     52     return transform(*parameters)
---> 54 return transform(parameters)

File /opt/monai/monai/transforms/io/dictionary.py:272, in SaveImaged.__call__(self, data)
    271     meta_data = d.get(meta_key) if meta_key is not None else None
--> 272     self.saver(img=d[key], meta_data=meta_data)
    273 return d

File /opt/monai/monai/transforms/io/array.py:408, in SaveImage.__call__(self, img, meta_data)
    407 meta_data = img.meta if isinstance(img, MetaTensor) else meta_data
--> 408 subject = meta_data[Key.FILENAME_OR_OBJ] if meta_data else str(self._data_index)
    409 patch_index = meta_data.get(Key.PATCH_INDEX, None) if meta_data else None

KeyError: 'filename_or_obj'

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
Input In [207], in <cell line: 5>()
     14 print(i)
     15 test_data['pred'] = sliding_window_inference(
     16     test_input,
     17     roi_size, sw_batch_size, net, sw_device= device, device="cpu")
---> 19 test_data = [post_transforms(i) for i in decollate_batch(test_data)]

Input In [207], in <listcomp>(.0)
     14 print(i)
     15 test_data['pred'] = sliding_window_inference(
     16     test_input,
     17     roi_size, sw_batch_size, net, sw_device= device, device="cpu")
---> 19 test_data = [post_transforms(i) for i in decollate_batch(test_data)]

File /opt/monai/monai/transforms/compose.py:173, in Compose.__call__(self, input_)
    171 def __call__(self, input_):
    172     for _transform in self.transforms:
--> 173         input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
    174     return input_

File /opt/monai/monai/transforms/transform.py:114, in apply_transform(transform, data, map_items, unpack_items, log_stats)
    112     else:
    113         _log_stats(data=data)
--> 114 raise RuntimeError(f"applying transform {transform}") from e

RuntimeError: applying transform <monai.transforms.io.dictionary.SaveImaged object at 0x7f22f00e5610>

```

A long thread here, and thanks for the help till now. Appreciate it a lot!

**Reply by KumoLiu** [reply]
Hi @lenlobo, can I know your MONAI version?

**Reply by lenlobo** [reply]
@KumoLiu the version would 0.9.1

**Reply by KumoLiu** [reply]
Hi @lenlobo, it seems there is nothing wrong with your part of the code, but could you please check the type of your `test_input` and `test_data['pred']` to make sure whether they are MetaTensor? If they are, also check whether `filename_or_obj ` is in `test_input.meta` and `test_data['pred'].meta`.
Hope it can help you, thanks!

**Reply by lenlobo** [reply]
@KumoLiu Hi, so, type of `test_input ` and `test_data['pred']` is `torch.Tensor` instead of MetaTensor, this was one of the differences I saw while trying to debug using the tutorials as a reference, that the outputs had a MetaTensor at the end of the tensor lists. Which I do not see in mine.

**Reply by KumoLiu** [reply]
Hi @lenlobo, can you use `DataLoader` in MONAI and check again your data type?
https://github.com/Project-MONAI/MONAI/blob/5b3b74143f9c8b985974e583b0a6f4029eef50e1/monai/data/dataloader.py#L21
```
first_sample = first(train_loader)
print("type: ", type(first_sample["image"]))
```

**Reply by lenlobo** [reply]
@KumoLiu I create a DataLoader class from the MONAI code you linked, and use that directly over torch.utils.data.DataLoader is what you are trying to say?

**Reply by KumoLiu** [reply]
Hi @lenlobo, from the code you post, I noticed that you are using `torch.utils.data.DataLoader`. I would suggest using `DataLoader` in MONAI directly. If you are creating a DataLoader by yourself, you can also replace `collate_fn` and `worker_init_fn` with MONAI components.
https://github.com/Project-MONAI/MONAI/blob/5b3b74143f9c8b985974e583b0a6f4029eef50e1/monai/data/dataloader.py#L84-L87

And I also noticed that you add `ToTensord ` in your transforms, you can just remove this line.
Hope it can help you, thanks!

**Reply by lenlobo** [reply]
Hi @KumoLiu, once I removed the ToTensord line, all my data did change from torch.Tensor to MetaTensor, now when I try to apply post transforms on my prediction by using `test_data = [post_transforms(i) for i in decollate_batch(test_data)]` I get these errors.
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File /opt/monai/monai/transforms/transform.py:90, in apply_transform(transform, data, map_items, unpack_items, log_stats)
     89         return [_apply_transform(transform, item, unpack_items) for item in data]
---> 90     return _apply_transform(transform, data, unpack_items)
     91 except Exception as e:

File /opt/monai/monai/transforms/transform.py:54, in _apply_transform(transform, parameters, unpack_parameters)
     52     return transform(*parameters)
---> 54 return transform(parameters)

File /opt/monai/monai/transforms/utility/dictionary.py:523, in ToTensord.inverse(self, data)
    522     # Remove the applied transform
--> 523     self.pop_transform(d, key)
    524 return d

File /opt/monai/monai/transforms/inverse.py:206, in TraceableTransform.pop_transform(self, data, key, check)
    192 """
    193 Return and pop the most recent transform.
    194 
   (...)
    204     - RuntimeError: data is neither `MetaTensor` nor dictionary
    205 """
--> 206 return self.get_most_recent_transform(data, key, check, pop=True)

File /opt/monai/monai/transforms/inverse.py:188, in TraceableTransform.get_most_recent_transform(self, data, key, check, pop)
    187 if check:
--> 188     self.check_transforms_match(all_transforms[-1])
    189 return all_transforms.pop() if pop else all_transforms[-1]

IndexError: list index out of range

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
File /opt/monai/monai/transforms/transform.py:90, in apply_transform(transform, data, map_items, unpack_items, log_stats)
     89         return [_apply_transform(transform, item, unpack_items) for item in data]
---> 90     return _apply_transform(transform, data, unpack_items)
     91 except Exception as e:

File /opt/monai/monai/transforms/transform.py:54, in _apply_transform(transform, parameters, unpack_parameters)
     52     return transform(*parameters)
---> 54 return transform(parameters)

File /opt/monai/monai/transforms/post/dictionary.py:664, in Invertd.__call__(self, data)
    663 with allow_missing_keys_mode(self.transform):  # type: ignore
--> 664     inverted = self.transform.inverse(input_dict)
    666 # save the inverted data

File /opt/monai/monai/transforms/compose.py:183, in Compose.inverse(self, data)
    182 for t in reversed(invertible_transforms):
--> 183     data = apply_transform(t.inverse, data, self.map_items, self.unpack_items, self.log_stats)
    184 return data

File /opt/monai/monai/transforms/transform.py:114, in apply_transform(transform, data, map_items, unpack_items, log_stats)
    113         _log_stats(data=data)
--> 114 raise RuntimeError(f"applying transform {transform}") from e

RuntimeError: applying transform <bound method ToTensord.inverse of <monai.transforms.utility.dictionary.ToTensord object at 0x7fada5dffd90>>

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
Input In [75], in <cell line: 1>()
----> 1 test_data = [post_transforms(i) for i in decollate_batch(test_data)]

Input In [75], in <listcomp>(.0)
----> 1 test_data = [post_transforms(i) for i in decollate_batch(test_data)]

File /opt/monai/monai/transforms/compose.py:173, in Compose.__call__(self, input_)
    171 def __call__(self, input_):
    172     for _transform in self.transforms:
--> 173         input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items, self.log_stats)
    174     return input_

File /opt/monai/monai/transforms/transform.py:114, in apply_transform(transform, data, map_items, unpack_items, log_stats)
    112     else:
    113         _log_stats(data=data)
--> 114 raise RuntimeError(f"applying transform {transform}") from e

RuntimeError: applying transform <monai.transforms.post.dictionary.Invertd object at 0x7fab21bc78b0>
```
Pretty close to fixing the issue here but need a bit more assistance! :) Thanks!

**Reply by KumoLiu** [reply]
Hi @lenlobo, if you remove `ToTensord` in test post transforms, you should also remove it in train post transforms. :) Thanks!

**Reply by lenlobo** [reply]
Hi @KumoLiu, you mean the transforms when I trained my model? Because I do not really have that code, I was using a model which is trained by someone else at my organization and just testing inputs by creating an inference script, or, are you talking about to_tensor=True in the post transforms mentioned by me in the above reply?

**Reply by lenlobo** [reply]
@KumoLiu Hi, this has kind off solved my issues, thank you so much! this helped a lot!
